elastic/apm-server
# otlp: Add beta support to receive logs

## Motivation/summary

Adds support to receive OTLP logs on the GRPC receiver. This feature is
_beta_ and should be used sparsely until the OTel spec declares it as
stable.

Currently, it's challenging to test this in the system tests since the
`opentelemetry-go` SDK doesn't expose any APIs for logs.

## Checklist




## How to test these changes

1. `docker-compose up -d`
2. `make`
3. `./apm-server -E output.elasticsearch.username=admin -E output.elasticsearch.password=changeme -e`
4. Run the following go program: (Make sure that you're using `go.opentelemetry.io/collector/model@v0.34.0`):
```go
package main

import (
	"context"
	"time"

	"go.opentelemetry.io/collector/model/otlpgrpc"
	"go.opentelemetry.io/collector/model/pdata"
	semconv "go.opentelemetry.io/collector/model/semconv/v1.5.0"
	"google.golang.org/grpc"
)

func main() {
	conn, err := grpc.Dial("localhost:8200", grpc.WithInsecure(), grpc.WithBlock())
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	client := otlpgrpc.NewLogsClient(conn)
	newLog := func(body interface{}) pdata.Logs {
		logs := pdata.NewLogs()
		resourceLogs := logs.ResourceLogs().AppendEmpty()
		logs.ResourceLogs().At(0).Resource().Attributes().InsertString(
			semconv.AttributeTelemetrySDKLanguage, "go",
		)
		instrumentationLogs := resourceLogs.InstrumentationLibraryLogs().AppendEmpty()
		otelLog := instrumentationLogs.Logs().AppendEmpty()
		otelLog.SetTraceID(pdata.NewTraceID([16]byte{1}))
		otelLog.SetSpanID(pdata.NewSpanID([8]byte{2}))
		otelLog.SetName("doOperation()")
		otelLog.SetSeverityNumber(pdata.SeverityNumberINFO)
		otelLog.SetSeverityText("Info")
		otelLog.SetTimestamp(pdata.NewTimestampFromTime(time.Now()))
		otelLog.Attributes().InsertString("key", "value")
		otelLog.Attributes().InsertDouble("key", 1234)

		switch b := body.(type) {
		case string:
			otelLog.Body().SetStringVal(b)
		case int:
			otelLog.Body().SetIntVal(int64(b))
		case float64:
			otelLog.Body().SetDoubleVal(float64(b))
		case bool:
			otelLog.Body().SetBoolVal(b)
		}
		return logs
	}

	_, err = client.Export(context.Background(), newLog("my log message"))
	if err != nil {
		panic(err)
	}
	_, err = client.Export(context.Background(), newLog(12334))
	if err != nil {
		panic(err)
	}
}
```
5. Install the APM Integration
6. Verify that the two log messages show up in the `Observability > Logs` UI.

<img width="1740" alt="image" src="https://user-images.githubusercontent.com/7286993/144819645-e13814b6-78d3-4b09-b5bb-d0257d60b553.png">

## Related issues

Closes #5491
